backend/project/routes_webauthn_duo.py

```diff
@@ -0,0 +1,282 @@
+
+    challenge = secrets.token_urlsafe(32)
+
+    # We strip the padding from the challenge stored in the session
+    # for the reasons outlined in the comment in webauthn_begin_activate.
+    session['challenge'] = challenge.rstrip('=')
+
+    webauthn_user = webauthn.WebAuthnUser(
+        user[2], # user.ukey
+        user[0], # user.username (or email in our case)
+        user[1], # user.display_name
+        user[3], # user.icon_url
+        user[4], # user.credential_id
+        user[5], # user.pub_key
+        user[6], # user.sign_count
+        user[7]) # user.rp_id
+
+    webauthn_assertion_options = webauthn.WebAuthnAssertionOptions(
+        webauthn_user, challenge)
+
+    return jsonify(webauthn_assertion_options.assertion_dict)
+
+
+@app.route('/verify_credential_info', methods=['POST'])
+def verify_credential_info():
+    challenge = session['challenge']
+    email = session['register_email']
+    display_name = session['register_display_name']
+    ukey = session['register_ukey']
+
+    registration_response = request.form
+    trust_anchor_dir = os.path.join(
+        os.path.dirname(os.path.abspath(__file__)), TRUST_ANCHOR_DIR)
+    trusted_attestation_cert_required = True
+    self_attestation_permitted = True
+    none_attestation_permitted = True
+
+    webauthn_registration_response = webauthn.WebAuthnRegistrationResponse(
+        RP_ID,
+        ORIGIN,
+        registration_response,
+        challenge,
+        trust_anchor_dir,
+        trusted_attestation_cert_required,
+        self_attestation_permitted,
+        none_attestation_permitted,
+        uv_required=False)  # User Verification
+
+    try:
+        webauthn_credential = webauthn_registration_response.verify()
+    except Exception as e:
+        return jsonify({'fail': 'Registration failed. Error: {}'.format(e)})
+
+    # Step 17.
+    #
+    # Check that the credentialId is not yet registered to any other user.
+    # If registration is requested for a credential that is already registered
+    # to a different user, the Relying Party SHOULD fail this registration
+    # ceremony, or it MAY decide to accept the registration, e.g. while deleting
+    # the older registration.
+    dbcursor.execute('SELECT credential_id FROM test WHERE credential_id = %s', (webauthn_credential.credential_id,))
+    credential_id_exists = dbcursor.fetchone()
+    
+    if credential_id_exists:
+        return make_response(
+            jsonify({
+                'fail': 'Credential ID already exists.'
+            }), 401)
+
+    dbcursor.execute('SELECT email FROM test WHERE email = %s', (email,))
+    existing_user = dbcursor.fetchone()
+    if not existing_user:
+        if sys.version_info >= (3, 0):
+            webauthn_credential.credential_id = str(
+                webauthn_credential.credential_id, "utf-8")
+            webauthn_credential.public_key = str(
+                webauthn_credential.public_key, "utf-8")
+        '''
+        user = User(
```

Comment:
I'm not sure I understand the purpose of the user class is (not saying it shouldn't exist, just that I'm dumb lol). I haven't used `Login_Manager` before, so I don't know what benefits this has over our current system other than exposing the `@login_required` route. Do you know what benefits there are?

Reply:
`User` is there just because `Flask-Login` needs some notion of a user class that implements certain methods (which, in our case, is taken care of by inheriting the default implementations from `UserMixin`). As far as benefits go, the documentation page probably explains better than I can: https://flask-login.readthedocs.io/en/latest/. Is it better than the existing system, though? I think it more or less achieves the same thing, but it certainly feels like a less hack-y solution.

Reply:
I think you're right. This is probably a more scalable solution than our hackier option with Flask sessions.